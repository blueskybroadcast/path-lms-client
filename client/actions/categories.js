import normalize from 'jsonapi-normalizer';

import { FETCH_CATEGORIES, DELETE_CATEGORY } from './types';

export const fetchCategories = () => async (dispatch, getState, api) => {
  const res = await api.get('/categories');

  dispatch({
    type: FETCH_CATEGORIES,
    payload: normalize(res.data)
  });
};

export const addCategory = category => async (dispatch, getState, api) => {
  const { name, parentId, hidden } = category;

  await api.post('/categories', {
    name,
    parent_id: parentId,
    hidden
  }).then(() => dispatch(fetchCategories()));
};

export const deleteCategory = id => async (dispatch, getState, api) => {
  await api.delete(`/categories/${id}`).then(() => {
    dispatch({
      type: DELETE_CATEGORY,
      payload: id
    });
  });
};

export const editCategory = (id, category) => async (dispatch, getState, api) => {
  const { name, parentId, hidden } = category;

  await api.put(`/categories/${id}`, {
    name,
    parent_id: parentId,
    hidden
  }).then(() => {
    dispatch(fetchCategories());
  });
};
