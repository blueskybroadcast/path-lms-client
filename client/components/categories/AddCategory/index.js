import AddCategory from './container';

export default AddCategory;
