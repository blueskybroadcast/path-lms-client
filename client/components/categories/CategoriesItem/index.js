import CategoriesItem from './container';

export default CategoriesItem;
