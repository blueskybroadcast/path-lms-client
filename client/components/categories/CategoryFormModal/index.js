import CategoryFormModal from './container';

export default CategoryFormModal;
