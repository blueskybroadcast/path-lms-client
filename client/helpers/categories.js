export const getParentListLevelOne = (categories) => {
  const parentsLevelOne = categories.filter(cat => cat.level === 0 && cat.childCount > 0);
  return parentsLevelOne.map(({ name, id }) => ({
    name,
    list: categories.filter(cat => cat.parentId === id)
  }));
};

export const getSortedCategories = (categoriesIds, categoriesData) => {
  const categories = [];

  categoriesIds.forEach((id) => {
    const data = categoriesData[id];
    if (!data.parent.id) {
      categories.push({
        id,
        level: 0,
        name: data.name,
        childCount: 0,
        hidden: data.hidden
      });
    }
  });

  const nextCategories = [...categories];
  let nextAddedCategoriesCount = 0;

  categories.forEach((cat, index) => {
    categoriesIds.forEach((id) => {
      const data = categoriesData[id];
      const parentId = data.parent.id;

      if (cat.id === parentId) {
        nextCategories.splice(
          index + 1 + nextAddedCategoriesCount,
          0,
          {
            id,
            level: 1,
            name: data.name,
            childCount: 0,
            parentId,
            hidden: data.hidden
          }
        );
        categories[index].childCount += 1;
        nextAddedCategoriesCount += 1;
      }
    });
  });

  const finalCategories = [...nextCategories];
  let finalAddedCategoriesCount = 0;

  nextCategories.forEach((cat, index) => {
    if (cat.level === 1) {
      categoriesIds.forEach((id) => {
        const data = categoriesData[id];
        if (cat.id === data.parent.id) {
          finalCategories.splice(
            index + 1 + finalAddedCategoriesCount,
            0,
            {
              id,
              level: 2,
              name: data.name,
              hidden: data.hidden
            }
          );
          nextCategories[index].childCount += 1;
          finalAddedCategoriesCount += 1;
        }
      });
    }
  });

  return finalCategories;
};

export const getCategoriesForSelect = (categoriesIds, categoriesData) => {
  if (!categoriesIds.length) return [];
  const remainingCategories = [...categoriesIds];
  const categories = [];

  categoriesIds.forEach((id) => {
    const data = categoriesData[id];
    if (!data.parent.id) {
      categories.push({
        id,
        name: data.name,
        hidden: data.hidden,
        children: []
      });
      const itemToRemoveIndex = remainingCategories.indexOf(id);
      remainingCategories.splice(itemToRemoveIndex, 1);
    }
  });

  const finalCategories = [...categories];

  categories.forEach((cat, index) => {
    remainingCategories.forEach((id) => {
      const data = categoriesData[id];
      const parentId = data.parent.id;

      if (cat.id === parentId) {
        const levelOneCat = {
          id,
          name: data.name,
          hidden: data.hidden,
          children: []
        };

        remainingCategories.forEach((childId) => {
          const childData = categoriesData[childId];
          const childParentId = childData.parent.id;

          if (id === childParentId) {
            const levelTwoCat = {
              id: childId,
              name: childData.name,
              hidden: childData.hidden,
              children: []
            };

            levelOneCat.children.push(levelTwoCat);
          }
        });

        finalCategories[index].children.push(levelOneCat);
      }
    });
  });

  return finalCategories;
};
