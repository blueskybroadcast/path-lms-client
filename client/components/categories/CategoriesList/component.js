import React from 'react';
import PropTypes from 'prop-types';

import CategoriesItem from '../CategoriesItem';

const CategoriesList = ({ loading, categoriesSortedList, categoriesData }) => (
  <section className="list">
    { loading ? (
      <div className="bsb-loader-container" data-role="courses-container">
        <div className="bsb-loader">
          <div className="loader-wrapper">
            <div className="spinner">
              <i className="icon icon-spinner icon-spin icon-large" />
            </div>
          </div>
        </div>
      </div>
    ) : (
      <ul>
        {categoriesSortedList && categoriesSortedList.map(({ id, level }) => (
          <CategoriesItem
            key={id}
            level={level}
            {...categoriesData[id]}
          />
        ))}
      </ul>
    )}
  </section>
);

CategoriesList.propTypes = {
  loading: PropTypes.bool.isRequired,
  categoriesSortedList: PropTypes.arrayOf(PropTypes.object).isRequired,
  categoriesData: PropTypes.objectOf(PropTypes.object)
};

CategoriesList.defaultProps = {
  categoriesData: {}
};

export default CategoriesList;
