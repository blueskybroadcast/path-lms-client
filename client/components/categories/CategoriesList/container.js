import { connect } from 'react-redux';

import { loadingCategoriesSelector } from '../../../selectors/loadingSelectors';
import {
  categoriesSortedIdsSelector, categoriesDataSelector
} from '../../../selectors/categoriesSelectors';

import CategoriesList from './component';

const mapStateToProps = state => ({
  loading: loadingCategoriesSelector(state),
  categoriesSortedList: categoriesSortedIdsSelector(state),
  categoriesData: categoriesDataSelector(state)
});

export default connect(mapStateToProps)(CategoriesList);
