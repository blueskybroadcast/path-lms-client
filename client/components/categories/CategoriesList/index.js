import CategoriesList from './container';

export default CategoriesList;
