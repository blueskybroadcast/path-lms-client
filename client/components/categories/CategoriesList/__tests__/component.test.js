import React from 'react';
import { shallow } from 'enzyme';
import CategoriesList from '../component';

import { categoriesSortedList, categoriesData } from '../../../../fixtures/categories';

test('should render CategoriesList in loading state correctly', () => {
  const wrapper = shallow(
    <CategoriesList
      loading
      categoriesSortedList={categoriesSortedList}
      categoriesData={categoriesData}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should render CategoriesList correctly', () => {
  const wrapper = shallow(
    <CategoriesList
      loading={false}
      categoriesSortedList={categoriesSortedList}
      categoriesData={categoriesData}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
