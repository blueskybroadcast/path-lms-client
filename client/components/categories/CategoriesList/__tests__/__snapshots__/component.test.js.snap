// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CategoriesList correctly 1`] = `
<section
  className="list"
>
  <ul>
    <Connect(CategoriesItem)
      courses={Array []}
      hidden={false}
      id="28"
      key="28"
      level={0}
      name="Another Level 0"
      parent={Object {}}
    />
    <Connect(CategoriesItem)
      courses={Array []}
      hidden={false}
      id="30"
      key="30"
      level={1}
      name="Another Level 0 child"
      parent={
        Object {
          "id": "28",
          "type": "category",
        }
      }
    />
    <Connect(CategoriesItem)
      courses={Array []}
      hidden={false}
      id="25"
      key="25"
      level={0}
      name="Level 0"
      parent={Object {}}
    />
    <Connect(CategoriesItem)
      courses={Array []}
      hidden={false}
      id="26"
      key="26"
      level={1}
      name="Level 0 child"
      parent={
        Object {
          "id": "25",
          "type": "category",
        }
      }
    />
    <Connect(CategoriesItem)
      courses={Array []}
      hidden={false}
      id="27"
      key="27"
      level={2}
      name="Level 0 child child"
      parent={
        Object {
          "id": "26",
          "type": "category",
        }
      }
    />
    <Connect(CategoriesItem)
      courses={Array []}
      hidden={false}
      id="29"
      key="29"
      level={0}
      name="Super Level 0"
      parent={Object {}}
    />
  </ul>
</section>
`;

exports[`should render CategoriesList in loading state correctly 1`] = `
<section
  className="list"
>
  <div
    className="bsb-loader-container"
    data-role="courses-container"
  >
    <div
      className="bsb-loader"
    >
      <div
        className="loader-wrapper"
      >
        <div
          className="spinner"
        >
          <i
            className="icon icon-spinner icon-spin icon-large"
          />
        </div>
      </div>
    </div>
  </div>
</section>
`;
